### PR TITLE
Extract trace context from `client_context`

### PR DESCRIPTION
**Pull request by Czechh**
### What does this PR do?
Adds a way of extracting trace data from `client_context` object - `custom` property. This will allow us to inject the trace data in other ways besides the payload's headers.

### Motivation
Distributed tracing - AWS-SDK Lambda invocation calls will be able to inject this data.

### Testing Guidelines

- New Unit tests
- Integration tests as regression.

### Types of Changes

- [ ] Bug fix
- [X] New feature
- [ ] Breaking change
- [ ] Misc (docs, refactoring, dependency upgrade, etc.)

### Check all that apply

- [X] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [X] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [X] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [X] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

File: tests/integration/snapshots/logs/sync-metrics_python37.log [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
+HTTP POST https://api.datadoghq.com/api/v1/distribution_points?api_key=XXXX Headers: ["Accept-Encoding:gzip, deflate", "Accept:*/*", "Connection:keep-alive", "Content-Length:460", "Content-Type:application/json", "User-Agent:datadogpy/0.38.0 (python 3.7.9; os linux; arch x86_64)", "x-datadog-parent-id:XXXX", "x-datadog-sampling-priority:2", "x-datadog-trace-id:XXXX"] Data: {"series": [{"metric": "hello.dog", "points": [[XXXX, [1.0]]], "type": "distribution", "host": null, "device": null, "tags": ["team:serverless", "role:hello", "dd_lambda_layer:datadog-python37_2.XX.0"], "interval": 10}, {"metric": "tests.integration.count", "points": [[XXXX, [21.0]]], "type": "distribution", "host": null, "device": null, "tags": ["test:integration", "role:hello", "dd_lambda_layer:datadog-python37_2.XX.0"], "interval": 10}]}
```
Comment (tianchu) [inline tests/integration/snapshots/logs/sync-metrics_python37.log:15]: Should mask the python version from the user agent? or at least mask the patch version?

File: datadog_lambda/tracing.py [hunk 2]
```diff
@@ -89,9 +91,30 @@ def _context_obj_to_headers(obj):
     }
 
 
-def extract_dd_trace_context(event):
+def get_dd_trace_data(low_headers, context):
```
Comment (tianchu) [inline datadog_lambda/tracing.py:94]: I don't think readers can really tell the differences between `get_dd_trace_data` and `extract_dd_trace_context` by just reading the function names. I think this suggests an actual problem and a better structure could look like this:

- extract_dd_trace_context(event, context)
  - extract_dd_trace_context_from_context(context)
  - extract_dd_trace_context_from_event(event)

File: scripts/run_integration_tests.sh [hunk 3]
```diff
@@ -10,7 +10,7 @@ set -e
 
 # These values need to be in sync with serverless.yml, where there needs to be a function
 # defined for every handler_runtime combination
-LAMBDA_HANDLERS=("async-metrics" "sync-metrics" "http-requests" "http-error")
```
Comment (tianchu) [inline scripts/run_integration_tests.sh:13]: why is this deleted?

Comment (Czechh) [reply scripts/run_integration_tests.sh:13]: I added it originally, after some time I think it's was a mistake my reasoning is the following.
1. It adds an external dependency that has failed before
2. It doesn't provide a lot of value and more overhead to the suite

File: datadog_lambda/tracing.py (hunk 2 above)
Comment (Czechh) [reply datadog_lambda/tracing.py:94]: Good call

File: tests/integration/snapshots/logs/async-metrics_python38_with_plugin.log [hunk 4]
```diff
@@ -2,23 +2,23 @@ START RequestId: XXXX Version: $LATEST
...
+{"traces": [[{"trace_id": "XXXX", "parent_id": "XXXX", "span_id": "XXXX", "service": "aws.lambda", "resource": "integration-plugin-dev-async-metrics_python38_with_plugin", "name": "aws.lambda", "error": 0, "start": XXXX, "duration": XXXX, "meta": {"_dd.origin": "lambda", "runtime-id": "XXXX", "cold_start": "false", "function_arn": "arn:aws:lambda:us-east-1:601427279990:function:integration-plugin-dev-async-metrics_python38_with_plugin", "function_version": "$LATEST", "request_id": "XXXX", "resource_names": "integration-plugin-dev-async-metrics_python38_with_plugin", "datadog_lambda": "2.23.0", "dd_trace": "0.41.2", "_dd.parent_source": "xray"}, "metrics": {"system.pid": XXXX, "_sampling_priority_v1": 2}, "type": "serverless"}]]}
```
Comment (nhinsch) [inline tests/integration/snapshots/logs/async-metrics_python38_with_plugin.log:21]: Seems like we should filter out the version numbers and replace them with "XXXX" so that the snapshots don't change every time we change the version number.

File: datadog_lambda/tracing.py [hunk 5]
```diff
@@ -89,9 +91,33 @@ def _context_obj_to_headers(obj):
...
+    parent_id = lowercase_headers.get(TraceHeader.PARENT_ID)
+    sampling_priority = lowercase_headers.get(TraceHeader.SAMPLING_PRIORITY)
+
+    return trace_id, parent_id, sampling_priority
+
+
+def extract_dd_trace_context(event, context):
     """
-    Extract Datadog trace context from the Lambda `event` object.
+    Extract Datadog trace context from the Lambda `event` object or
```
Comment (nhinsch) [inline datadog_lambda/tracing.py:119]: Nit: I would just say extract from the `event` or `context` objects. Technically `client_context` is a key within the context object.

File: tests/integration/snapshots/logs/async-metrics_python38_with_plugin.log (hunk 4 above)
Comment (tianchu) [reply tests/integration/snapshots/logs/async-metrics_python38_with_plugin.log:21]: Yeah, agreed, we should probably mask them. I believe @meghna-dd added these very recently.

Comment (Czechh) [reply tests/integration/snapshots/logs/async-metrics_python38_with_plugin.log]: Done!

File: datadog_lambda/tracing.py (hunk 5 above)
Comment (Czechh) [reply datadog_lambda/tracing.py:119]: Done!

**Review by tianchu — APPROVED** [review]

**Comment by tianchu** [comment]
Do you have the PR to inject trace context to client_context? Can you link it here for context if you do?

**Comment by Czechh** [comment]
@tianchu https://github.com/DataDog/dd-trace-py/pull/1748
